Allow custom priority threshold

lixiang233 opened the pull request:
Fixes #329 
Introduce a new configuration for each strategies which allows users to define a PriorityClass which is used by PodEvictor to check if a pod is evictable.

TODO:
- [x] Add PriorityClass to `StrategyParameters` and read and set it to PodEvictor
- [x] e2e tests for priority class
- [x] README
- [x] e2e tests for priority

lixiang233 commented:
/cc @ingvagabund  @seanmalloy 
/assign @damemi 
Please take a look, this pr is ready for review and will be ready for merge once I have updated README.

damemi commented on pkg/descheduler/strategies/duplicates.go:
```diff
@@ -44,6 +44,17 @@ func RemoveDuplicatePods(
+		klog.V(1).Infof("Error setting priority threshold, %#v", err)
+		return
+	}
+	klog.V(1).Infof("Running RemoveDuplicatePods with priority threshold %d", priority)
```
I think we should only log the threshold if it's being set to something that's not default.

damemi commented on pkg/descheduler/strategies/lownodeutilization.go:
```diff
@@ -53,6 +53,17 @@ const (
 // LowNodeUtilization evicts pods from overutilized nodes to underutilized nodes. Note that CPU/Memory requests are used
 // to calculate nodes' utilization and not the actual resource usage.
 func LowNodeUtilization(ctx context.Context, client clientset.Interface, strategy api.DeschedulerStrategy, nodes []*v1.Node, podEvictor *evictions.PodEvictor) {
+	var priorityClass string
+	if strategy.Params != nil {
+		priorityClass = strategy.Params.ThresholdPriorityClassName
+	}
+	priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, priorityClass)
+	if err != nil {
+		klog.V(1).Infof("Error setting priority threshold, %#v", err)
+		return
+	}
```
You could simplify this a bit like:
```go
if strategy.Params != nil {
  priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, strategy.Params.ThresholdPriorityClassName)
  if err != nil {
    klog.V(1).Errorf("Error setting priority threshold: %+v", err)
    return
  }
  klog.V(1).Infof("Running LowNodeUtilization with priority threshold %d", priority)
}
```

Or just pass `*strategyParams` to `podEvictor.SetPriorityThresholdFromPriorityClass` and do the nil check there (and the log message below this if it's successfully set). That would clean up some of the copy paste in each strategy

damemi commented on pkg/descheduler/strategies/pod_lifetime.go:
```diff
@@ -49,6 +49,13 @@ func PodLifeTime(ctx context.Context, client clientset.Interface, strategy api.D
 		klog.V(1).Info(err)
 		return
 	}
+	priorityClass := strategy.Params.ThresholdPriorityClassName
+	priority, err := podEvictor.SetPriorityThresholdFromPriorityClass(ctx, client, priorityClass)
```
any reason to not pass `strategy.Params.ThresholdPriorityClassName` directly to the function?

damemi commented:
btw when your PR is ready for review in the future just remove the [WIP] and we'll know it's ready :)
We can /hold if you're waiting for pieces like README updates that will depend on feedback
/retitle Allow custom priority threshold
/hold

lixiang233 commented:
> We could also consider just adding this as a `filter` function 

@damemi  I prefer this approach, as `IsPodEvictableBasedOnPriority` will be called for every pod no matter it's in `IsEvictable` or the filter function, just moving `IsPodEvictableBasedOnPriority` out of `IsEvictable` won't affect the performance. and if we add it to `EvictPod`, we may do some useless calculations because some pods may not be evictable and we could filter out them before running the strategy.

At first I've thought of many ways to do this, given we'll set priority threshold every time a strategy starts and our strategies are running one by one, I think it doesn't matter if one strategy may affect others, besides I think we should do all `IsEvictable` checks together in one function, so I choose to make the threshold a param of `PodEvictor`. That's not a very good implementation, I'll reimplement it using the methods mentioned above and add another configuration to set the threshold value directly.

ingvagabund commented:
I second what @damemi says.
For strategies which are not cluster wide (RemovePodsViolatingInterPodAntiAffinity, RemovePodsViolatingNodeAffinity, RemovePodsViolatingNodeTaints, ...), you can extend the filters so you don't bother with pods what have higher priority class than ThresholdPriorityClassName. In which case the pod evictor does not have to check the priority class at all.
For the remaining strategies (RemoveDuplicates, LowNodeUtilization), you can always check a pod's class right before you pass the pod to the evictor. So the pod evictor can stay intact.

lixiang233 commented:
As `All types of pods with the annotation descheduler.alpha.kubernetes.io/evict are evicted.` and this annotation is checked in `IsEvictable`, if I add the priority check as an individual filter function, and use them like:
```
podutil.WithFilter(func(pod *v1.Pod) bool {
      return podEvictor.IsEvictable(pod) && isPodEvictableByPriority(pod)
})
```
Although a pod has the annotation, if its priority is above priority threshold, it won't be evictable.

So if we want this annotation still work for priority filtering, I have some solutions:
1.add priority check to `IsEvictable` and refactor `IsEvictable` to have a parameter.
2.check this annotation again in `isPodEvictableByPriority`
3.make annotation check another filter func and call them like:
```
podutil.WithFilter(func(pod *v1.Pod) bool {
      return (podEvictor.IsEvictable(pod) && isPodEvictableByPriority(pod)) || isPodEvictableByAnnotation(pod)
})
```
I prefer 1, what do you think @damemi  @ingvagabund ?

lixiang233 commented:
I have implemented this
> 1.add priority check to `IsEvictable` and refactor `IsEvictable` to have a parameter.

and added `ThresholdPriority` to strategy's params.

ingvagabund commented on pkg/utils/pod.go:
```diff
@@ -112,15 +112,12 @@ func IsCriticalPod(pod *v1.Pod) bool {
 	if IsMirrorPod(pod) {
 		return true
 	}
-	if pod.Spec.Priority != nil && IsCriticalPodBasedOnPriority(*pod.Spec.Priority) {
```
This needs to stay here. You should never be allowed to evict critical pods.

damemi replied:
Isn't this check basically replaced by IsPodEvictableBasedOnPriority?

ingvagabund replied:
Given the way it's implemented:

```
func IsPodEvictableBasedOnPriority(pod *v1.Pod, priority int32) bool {
	return pod.Spec.Priority == nil || *pod.Spec.Priority < priority
}
```

a user can choose any int32 value. Even the one greater than the system critical priority. Thus, if improperly configured the descheduler can evict even the critical pods. `return pod.Spec.Priority == nil || (*pod.Spec.Priority < priority && *pod.Spec.Priority < SystemCriticalPriority)` would do a bit better.

My point is `IsEvictable` function is supposed to be static. The hardline for deciding when a pod is evictable or not without any degree of flexibility. If `IsEvictable` decides a pod is not evictable, it is not. It's a safe guard condition that can not be changed from the outside (except for setting some true/false flags).

damemi replied:
Good points, I did notice this change at first too but thought it was just being covered by the new code.

In that case, we should probably keep it the same as it was right?

damemi replied:
@lixiang233 what do you think?

lixiang233 replied:
Agree, if someone wants to evict system-critical pods, he can add `descheduler.alpha.kubernetes.io/evict` annotation. I'll add `*pod.Spec.Priority < SystemCriticalPriority` to `IsPodEvictableBasedOnPriority`.

lixiang233 commented:
I've moved `IsPodEvictableBasedOnPriority` to the `evictions` package and added e2e testcase for `ThresholdPriority`.

please take a look @damemi  @ingvagabund

lixiang233 commented:
I've completed the change mentioned before and added a validation to the threshold. @damemi  @ingvagabund

seanmalloy commented on pkg/utils/priority.go:
```diff
@@ -33,3 +39,37 @@ func PodMatchesTermsNamespaceAndSelector(pod *v1.Pod, namespaces sets.String, se
+		priorityClass, err := client.SchedulingV1().PriorityClasses().Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return 0, err
+		} else {
+			return priorityClass.Value, nil
+		}
```
```suggestion
		}
		return priorityClass.Value, nil
```

lixiang233 replied:
done.

ingvagabund commented:
/lgtm

ingvagabund commented:
/hold cancel
